File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
#     answer = 0
#     visited= [False]*(len(words))
#     if target in words:
#         q= deque([(begin,0)])
        
#         while q:
#             cur_word, step = q.popleft()
            
#             if cur_word == target:
#                 return step
            
#             for i in range(0,len(words)):
#                 if not visited[i] :
#                     differ_cnt =0
#                     for j in range(0,len(words[i])):
#                         if cur_word[j]!= words[i][j]:
#                             differ_cnt+=1
                            
                    
#                     if differ_cnt==1:
#                         visited[i]= True
#                         q.append((words[i],step+1))    
#                     else: continue
    
    #else: return 0
    answer = 0
    #한번에 한개의 알파벳,
    #words에 있는 단어로만 변환 가능
    n = len(words)
    visited = [False] *(n)
    if target not in words :
        return 0
    else:
        q= deque ([(begin, 0)])
        while q :
            cur_word, step = q.popleft()
            if cur_word == target :
                return step
            
            for i in range (0, len(words)) :
                if visited[i] :
                    continue
                    
                next_word = words[i]
                diff = 0
                for j in range (0, len(cur_word)):
                    if next_word[j] != cur_word[j] :
                        diff +=1
                
                if diff >1 :
                    continue
                else:
                    visited[i] = True
                    q.append((next_word, step+1))
    
    
                
            

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        
        ArrayDeque <String> q = new ArrayDeque<>();
        HashMap <String, Integer> hm = new HashMap<>();
        int n = words.length;
        boolean visited[] = new boolean[n];
        
        hm.put(begin, 0);
        q.offer(begin);
        
        
        while (!q.isEmpty()) {
            String cur_str = q.poll();
            
            if (cur_str.equals(target)) {
                break;
            }
            
            for (int i =0; i< words.length; i++) {
                String next_str = words[i];
                if (is_one_diff(cur_str, next_str)) {
                    if (!visited[i]) {
                        q.offer(next_str);
                        hm.put(next_str, hm.get(cur_str) +1);
                        visited[i] = true;
                    }
                }
            }
        }
        
        
        if (hm.containsKey(target)) {
            answer = hm.get(target);
        }else {
            return 0;
        }
        
    
        return answer;
    }
    
    private boolean is_one_diff (String str1, String str2) {
        char [] arr1 = str1.toCharArray();
        char [] arr2 = str2.toCharArray();
        int cnt = 0;
        for (int i =0; i< arr1.length; i++) {
            if (arr1[i] != arr2[i]) {
                cnt +=1;
            }
        }
        
        if (cnt >1) {
            return false;
        }
        if (cnt <=0) {
            return false;
        }
        
        return true;
    }
    
}